File: src/harness/context/UnifiedContextBuilder.ts
/**
 * 统一上下文构建器
 *
 * 【架构定位】
 * 上下文系统的统一入口，整合所有上下文相关组件
 * 解决组件分散、调用复杂的问题
 *
 * 【设计原则】
 * - 统一入口：一个类，一个方法，构建完整上下文
 * - 内部协调：自动协调各组件的工作顺序和依赖关系
 * - 向后兼容：不修改原有组件，只做整合
 * - 性能优化：缓存常用片段，减少重复计算
 *
 * 【整合的组件】
 * - ConstitutionPromptBuilder：宪法级系统 Prompt
 * - UnifiedContextPipeline：统一上下文管道（场景、情感、记忆、用户画像）
 * - LLMContextBuilder：智能记忆筛选
 * - ContextWindowManager：上下文窗口管理
 * - ContextFileRegistry：项目文件上下文
 * - ContextReferenceResolver：@引用解析
 *
 * 【使用方式】
 * const builder = UnifiedContextBuilder.getInstance();
 * const context = await builder.buildContext({
 *   input: userInput,
 *   userId: 'user123',
 *   includeSystemPrompt: true,
 *   includeMemory: true,
 * });
 *
 * @module UnifiedContextBuilder
 * @version 0.1.0
 * @status Alpha - 框架实现，功能待完善
 * @warning 生产环境慎用，API 可能有变更
 * @since 2026-06-24
 */

import { Logger } from '../../utils/Logger';
import type { ChatMessage, UserInput } from '../types';
import type { ConstitutionPromptBuilder } from '../../core/ConstitutionPromptBuilder';
import type {
  UnifiedContextPipeline,
  UnifiedContext,
} from '../../core/UnifiedContextPipeline';
import type { ContextWindowManager } from './ContextWindowManager';

// ========== 常量定义 ==========

/**
 * 上下文构建选项
 */
export interface ContextBuildOptions {
  /** 用户输入 */
  input: UserInput;
  /** 用户 ID */
  userId?: string;
  /** 是否包含系统 Prompt */
  includeSystemPrompt?: boolean;
  /** 是否包含记忆 */
  includeMemory?: boolean;
  /** 是否包含项目文件上下文 */
  includeFileContext?: boolean;
  /** 是否解析 @引用 */
  resolveReferences?: boolean;
  /** 最大 Token 预算 */
  maxTokens?: number;
  /** 场景类型 */
  scene?: string;
}

/**
 * 上下文构建状态
 */
export type ContextBuildStatus = 'success' | 'partial' | 'failed';

/**
 * 上下文构建结果
 */
export interface ContextBuildResult {
  /** 完整的消息列表 */
  messages: ChatMessage[];
  /** 系统 Prompt */
  systemPrompt?: string;
  /** 记忆内容 */
  memories?: string[];
  /** 上下文统计 */
  stats: ContextStats;
  /** 构建状态 */
  status: ContextBuildStatus;
  /** 错误详情（如果有） */
  errors?: Array<{
    component: string;
    message: string;
  }>;
}

/**
 * 上下文统计
 */
export interface ContextStats {
  /** 总消息数 */
  totalMessages: number;
  /** 估算 Token 数 */
  estimatedTokens: number;
  /** 记忆数量 */
  memoryCount: number;
  /** 文件上下文数量 */
  fileContextCount: number;
  /** 引用解析数 */
  referenceCount: number;
  /** 构建耗时（ms） */
  buildTime: number;
}

/**
 * 统一上下文构建器
 *
 * 【注意】
 * 这是统一入口类，内部协调各个上下文组件。
 * 当前版本为基础实现，后续会逐步完善。
 */
export class UnifiedContextBuilder {
  private static instance: UnifiedContextBuilder;

  /** 是否启用缓存 */
  private cacheEnabled = true;

  /** 系统 Prompt 缓存 */
  private systemPromptCache: string | null = null;

  /** 缓存过期时间（ms） */
  private readonly cacheTTL = 5 * 60 * 1000; // 5 分钟

  /** 缓存时间戳 */
  private cacheTimestamp = 0;

  /** 构建计数 */
  private buildCount = 0;

  /** 总耗时 */
  private totalBuildTime = 0;

  private constructor() {
    Logger.info('🧩 统一上下文构建器已初始化', 'ContextBuilder');
  }

  static getInstance(): UnifiedContextBuilder {
    if (!UnifiedContextBuilder.instance) {
      UnifiedContextBuilder.instance = new UnifiedContextBuilder();
    }
    return UnifiedContextBuilder.instance;
  }

  /**
   * 重置单例实例（测试用）
   *
   * 【注意】
   * - 仅供测试使用，生产环境请勿调用
   * - 会清除所有缓存和统计数据
   * - 调用后下次 getInstance() 会创建新实例
   */
  static resetInstance(): void {
    if (UnifiedContextBuilder.instance) {
      UnifiedContextBuilder.instance = null as any;
    }
  }

  /**
   * 创建测试用独立实例（测试用）
   *
   * 【注意】
   * - 仅供测试使用，生产环境请勿调用
   * - 创建的是独立实例，不影响单例
   */
  static createTestInstance(): UnifiedContextBuilder {
    return new UnifiedContextBuilder();
  }

  /**
   * 构建完整上下文
   *
   * @param options 构建选项
   * @returns 上下文构建结果
   *
   * 【错误降级设计】
   * - 每个组件独立 try-catch 隔离
   * - 单个组件失败不影响整体构建
   * - 失败时跳过该组件，继续构建其他部分
   * - 记录详细的错误信息和降级策略
   * - 返回状态标记：success / partial / failed
   */
  async buildContext(
    options: ContextBuildOptions
  ): Promise<ContextBuildResult> {
    const startTime = Date.now();
    this.buildCount++;

    Logger.debug(
      `开始构建上下文 (用户: ${options.userId || 'anonymous'}, 场景: ${options.scene || 'default'})`,
      'ContextBuilder'
    );

    const messages: ChatMessage[] = [];
    let systemPrompt: string | undefined;
    let memories: string[] = [];
    const errors: Array<{ component: string; message: string }> = [];
    let failedComponents = 0;
    let totalComponents = 0;

    // ========== 1. 构建系统 Prompt ==========
    if (options.includeSystemPrompt !== false) {
      totalComponents++;
      try {
        systemPrompt = await this.buildSystemPrompt(options);
        if (systemPrompt) {
          messages.push({
            role: 'system',
            content: systemPrompt,
          });
        }
        Logger.debug('系统 Prompt 构建成功', 'ContextBuilder');
      } catch (error) {
        failedComponents++;
        const errorMsg = error instanceof Error ? error.message : String(error);
        errors.push({
          component: 'SystemPrompt',
          message: errorMsg,
        });
        Logger.warn(
          `系统 Prompt 构建失败，已跳过: ${errorMsg}`,
          'ContextBuilder'
        );
      }
    }

    // ========== 2. 解析 @引用 ==========
    if (options.resolveReferences !== false) {
      totalComponents++;
      try {
        // TODO: 集成 ContextReferenceResolver
        // const references = await this.resolveReferences(options);
        Logger.debug('引用解析: 功能待实现', 'ContextBuilder');
      } catch (error) {
        failedComponents++;
        const errorMsg = error instanceof Error ? error.message : String(error);
        errors.push({
          component: 'ReferenceResolver',
          message: errorMsg,
        });
        Logger.warn(`引用解析失败，已跳过: ${errorMsg}`, 'ContextBuilder');
      }
    }

    // ========== 3. 加载记忆 ==========
    if (options.includeMemory !== false) {
      totalComponents++;
      try {
        memories = await this.loadMemories(options);
        // TODO: 将记忆注入到上下文中
        Logger.debug(`记忆加载完成: ${memories.length} 条`, 'ContextBuilder');
      } catch (error) {
        failedComponents++;
        const errorMsg = error instanceof Error ? error.message : String(error);
        errors.push({
          component: 'MemoryLoader',
          message: errorMsg,
        });
        Logger.warn(`记忆加载失败，已跳过: ${errorMsg}`, 'ContextBuilder');
      }
    }

    // ========== 4. 加载项目文件上下文 ==========
    if (options.includeFileContext !== false) {
      totalComponents++;
      try {
        // TODO: 集成 ContextFileRegistry
        // const fileContext = await this.loadFileContext(options);
        Logger.debug('文件上下文: 功能待实现', 'ContextBuilder');
      } catch (error) {
        failedComponents++;
        const errorMsg = error instanceof Error ? error.message : String(error);
        errors.push({
          component: 'FileContext',
          message: errorMsg,
        });
        Logger.warn(
          `文件上下文加载失败，已跳过: ${errorMsg}`,
          'ContextBuilder'
        );
      }
    }

    // ========== 5. 应用窗口管理 ==========
    totalComponents++;
    try {
      if (this.windowManager && messages.length > 0) {
        const managedMessages = this.windowManager.manageWindow(messages);
        messages.length = 0;
        messages.push(...managedMessages);
        Logger.debug(
          `窗口管理完成: ${managedMessages.length} 条消息 (原 ${messages.length} 条)`,
          'ContextBuilder'
        );
      } else if (!this.windowManager) {
        Logger.debug(
          '窗口管理: ContextWindowManager 未注入，跳过',
          'ContextBuilder'
        );
      }
    } catch (error) {
      failedComponents++;
      const errorMsg = error instanceof Error ? error.message : String(error);
      errors.push({
        component: 'WindowManager',
        message: errorMsg,
      });
      Logger.warn(`窗口管理失败，已跳过: ${errorMsg}`, 'ContextBuilder');
    }

    // ========== 计算构建状态 ==========
    let status: ContextBuildStatus = 'success';
    if (failedComponents > 0 && failedComponents < totalComponents) {
      status = 'partial';
    } else if (failedComponents === totalComponents && totalComponents > 0) {
      status = 'failed';
    }

    const buildTime = Date.now() - startTime;
    this.totalBuildTime += buildTime;

    const stats: ContextStats = {
      totalMessages: messages.length,
      estimatedTokens: this.estimateTokens(messages),
      memoryCount: memories.length,
      fileContextCount: 0, // TODO: 实际统计
      referenceCount: 0, // TODO: 实际统计
      buildTime,
    };

    Logger.debug(
      `上下文构建完成: 状态=${status}, ${messages.length} 条消息, 约 ${stats.estimatedTokens} tokens, 耗时 ${buildTime}ms`,
      'ContextBuilder'
    );

    if (errors.length > 0) {
      Logger.debug(
        `组件失败数: ${failedComponents}/${totalComponents}`,
        'ContextBuilder'
      );
    }

    return {
      messages,
      systemPrompt,
      memories,
      stats,
      status,
      errors: errors.length > 0 ? errors : undefined,
    };
  }

  /**
   * 构建系统 Prompt
   *
   * 优先使用注入的 ConstitutionPromptBuilder（完整宪法级 Prompt）
   * 降级到内置基础 Prompt（当 Component 未注入时）
   */
  private async buildSystemPrompt(
    options: ContextBuildOptions
  ): Promise<string> {
    // 检查缓存
    if (
      this.cacheEnabled &&
      this.systemPromptCache &&
      Date.now() - this.cacheTimestamp < this.cacheTTL
    ) {
      Logger.debug('使用缓存的系统 Prompt', 'ContextBuilder');
      return this.systemPromptCache;
    }

    let systemPrompt: string;

    // 尝试使用集成的 ConstitutionPromptBuilder
    if (this.constitutionPromptBuilder) {
      try {
        systemPrompt =
          await this.constitutionPromptBuilder.buildConstitutionPrompt(
            options.userId
          );
        Logger.debug(
          '使用 ConstitutionPromptBuilder 生成系统 Prompt',
          'ContextBuilder'
        );
      } catch (error) {
        const errorMsg = error instanceof Error ? error.message : String(error);
        Logger.warn(
          `ConstitutionPromptBuilder 调用失败，降级到基础 Prompt: ${errorMsg}`,
          'ContextBuilder'
        );
        systemPrompt = this.generateBasicSystemPrompt(options);
      }
    } else {
      // 降级：使用内置基础 Prompt
      Logger.debug(
        'ConstitutionPromptBuilder 未注入，使用基础系统 Prompt',
        'ContextBuilder'
      );
      systemPrompt = this.generateBasicSystemPrompt(options);
    }

    // 更新缓存
    if (this.cacheEnabled) {
      this.systemPromptCache = systemPrompt;
      this.cacheTimestamp = Date.now();
    }

    return systemPrompt;
  }

  /**
   * 生成基础系统 Prompt
   */
  private generateBasicSystemPrompt(_options: ContextBuildOptions): string {
    // 简化实现，后续替换为 ConstitutionPromptBuilder
    return `你是家百星（Jiabaixing），一个智能助手。

## 核心原则
- 保持专业、友好、高效的沟通风格
- 优先使用工具来获取准确信息
- 遇到不确定的问题，诚实说明而不是编造
- 保护用户隐私和数据安全

## 能力范围
- 代码开发与调试
- 文件操作与管理
- 网络搜索与信息获取
- 系统命令执行
- 多模态内容理解

## 行为准则
- 仔细思考后再行动
- 每一步都要验证结果
- 出错时及时调整策略
- 从错误中学习改进`;
  }

  /**
   * 加载记忆
   *
   * 优先使用注入的 UnifiedContextPipeline（完整上下文管道）
   * 降级到空数组（当 Pipeline 未注入时）
   */
  private async loadMemories(options: ContextBuildOptions): Promise<string[]> {
    if (this.contextPipeline && options.userId) {
      try {
        const pipelineContext: UnifiedContext =
          await this.contextPipeline.buildContext(
            options.input.text || '',
            options.userId
          );

        const memories: string[] = [];

        // 提取场景上下文
        if (pipelineContext.scene) {
          memories.push(`[场景] ${pipelineContext.scene}`);
        }

        // 提取情感状态
        if (pipelineContext.emotion) {
          memories.push(`[情感] ${pipelineContext.emotion}`);
        }

        // 提取记忆内容
        if (pipelineContext.memories && pipelineContext.memories.length > 0) {
          const memoryTexts = pipelineContext.memories
            .map(
              (m: {
                content?: string;
                type?: string;
                relevanceScore?: number;
              }) => {
                const prefix = m.type ? `[${m.type}] ` : '';
                return `${prefix}${m.content || ''}`;
              }
            )
            .filter((t: string) => t.length > 0);
          memories.push(...memoryTexts);
        }

        // 提取用户画像
        if (pipelineContext.userProfile) {
          const profile = pipelineContext.userProfile as Record<
            string,
            unknown
          >;
          const profileEntries = Object.entries(profile)
            .filter(([, v]) => v !== undefined && v !== null)
            .map(([k, v]) => `[画像:${k}] ${String(v)}`);
          memories.push(...profileEntries);
        }

        Logger.debug(
          `使用 UnifiedContextPipeline 加载 ${memories.length} 条记忆上下文`,
          'ContextBuilder'
        );

        return memories;
      } catch (error) {
        const errorMsg = error instanceof Error ? error.message : String(error);
        Logger.warn(
          `UnifiedContextPipeline 调用失败，降级到空记忆: ${errorMsg}`,
          'ContextBuilder'
        );
        return [];
      }
    }

    // 降级：无 Pipeline 或 userId 时的行为
    Logger.debug(
      '记忆加载: UnifiedContextPipeline 未注入或无 userId，返回空记忆',
      'ContextBuilder'
    );

    return [];
  }

  /**
   * 估算 Token 数
   */
  private estimateTokens(messages: ChatMessage[]): number {
    // 简化估算：中文 1.5 字符/token，英文 4 字符/token
    let totalChars = 0;
    for (const msg of messages) {
      totalChars += msg.content?.length || 0;
      totalChars += (msg.role?.length || 0) + 4; // role + 分隔符
    }

    // 简单估算：按平均 2 字符/token
    return Math.ceil(totalChars / 2);
  }

  /**
   * 获取构建统计
   */
  getBuildStats(): {
    totalBuilds: number;
    averageBuildTime: number;
    cacheHitRate: number;
  } {
    return {
      totalBuilds: this.buildCount,
      averageBuildTime:
        this.buildCount > 0 ? this.totalBuildTime / this.buildCount : 0,
      cacheHitRate: 0, // TODO: 实际统计
    };
  }

  /**
   * 清除缓存
   */
  clearCache(): void {
    this.systemPromptCache = null;
    this.cacheTimestamp = 0;
    Logger.info('上下文缓存已清除', 'ContextBuilder');
  }

  /**
   * 启用/禁用缓存
   */
  setCacheEnabled(enabled: boolean): void {
    this.cacheEnabled = enabled;
    Logger.info(`上下文缓存已${enabled ? '启用' : '禁用'}`, 'ContextBuilder');
  }

  /**
   * 检查缓存是否启用
   */
  isCacheEnabled(): boolean {
    return this.cacheEnabled;
  }

  // ========== 组件引用（依赖注入） ==========

  /** 宪法级系统 Prompt 构建器 */
  private constitutionPromptBuilder: ConstitutionPromptBuilder | null = null;

  /** 统一上下文管道 */
  private contextPipeline: UnifiedContextPipeline | null = null;

  /** 上下文窗口管理器 */
  private windowManager: ContextWindowManager | null = null;

  /**
   * 注入 ConstitutionPromptBuilder
   * 用于生成符合宪法原则的系统 Prompt
   */
  setConstitutionPromptBuilder(builder: ConstitutionPromptBuilder): void {
    this.constitutionPromptBuilder = builder;
    this.clearCache(); // 清除旧的系统 Prompt 缓存
    Logger.info('ConstitutionPromptBuilder 已注入', 'ContextBuilder');
  }

  /**
   * 注入 UnifiedContextPipeline
   * 用于加载场景上下文、情感状态、记忆和用户画像
   */
  setContextPipeline(pipeline: UnifiedContextPipeline): void {
    this.contextPipeline = pipeline;
    Logger.info('UnifiedContextPipeline 已注入', 'ContextBuilder');
  }

  /**
   * 注入 ContextWindowManager
   * 用于管理上下文 Token 窗口，超阈值时自动压缩
   */
  setWindowManager(manager: ContextWindowManager): void {
    this.windowManager = manager;
    Logger.info('ContextWindowManager 已注入', 'ContextBuilder');
  }

  /**
   * 检查所有关键组件是否已注入
   */
  isFullyIntegrated(): boolean {
    return (
      this.constitutionPromptBuilder !== null &&
      this.contextPipeline !== null &&
      this.windowManager !== null
    );
  }

  /**
   * 获取组件注入状态
   */
  getIntegrationStatus(): Record<string, boolean> {
    return {
      constitutionPromptBuilder: this.constitutionPromptBuilder !== null,
      contextPipeline: this.contextPipeline !== null,
      windowManager: this.windowManager !== null,
    };
  }
}

export default UnifiedContextBuilder;

File: python/agent/context/unified_context_builder.py
"""统一上下文构建器（UnifiedContextBuilder）。

对标 TS 侧 src/harness/context/UnifiedContextBuilder.ts。

作为上下文系统的统一入口，整合所有上下文相关组件：
- UnifiedContextPipeline：上下文数据生成（场景、情感、记忆、画像）
- UnifiedContextOrchestrator：组件编排和执行
- LLMContextBuilder：智能记忆筛选

提供简洁的 build_context(options) 接口，内部协调所有组件的协作。

Usage:
    builder = UnifiedContextBuilder()
    result = await builder.build_context(
        user_input="帮我写一个排序函数",
        user_id="user_1",
        include_system_prompt=True,
        include_memory=True,
    )
    logger.info("状态: {result.status}, 消息数: {len(result.messages)}")
"""

from __future__ import annotations

import time
from dataclasses import dataclass, field
from typing import Any

from agent.context.llm_context_builder import LLMContextBuilder, LLMContextBuilderConfig
from agent.context.unified_context_pipeline import (
    UnifiedContext,
    UnifiedContextPipeline,
)
import logging
logger = logging.getLogger(__name__)


@dataclass
class ContextBuildOptions:
    """上下文构建选项。

    Attributes:
        user_input: 用户输入。
        user_id: 用户ID。
        include_system_prompt: 是否包含系统 Prompt。
        include_memory: 是否包含记忆。
        include_file_context: 是否包含文件上下文。
        resolve_references: 是否解析 @引用。
        max_tokens: 最大 Token 预算。
        scene: 场景类型（不指定则自动检测）。
        history: 历史消息列表。
        history_limit: 历史消息数量限制。
    """

    user_input: str = ""
    user_id: str = ""
    include_system_prompt: bool = True
    include_memory: bool = True
    include_file_context: bool = False
    resolve_references: bool = False
    max_tokens: int = 4096
    scene: str = ""
    history: list[dict[str, str]] = field(default_factory=list)
    history_limit: int = 20


@dataclass
class ContextStats:
    """上下文统计。

    Attributes:
        total_messages: 总消息数。
        estimated_tokens: 估算 Token 数。
        memory_count: 记忆数量。
        file_context_count: 文件上下文数量。
        reference_count: 引用解析数。
        build_time_ms: 构建耗时（毫秒）。
    """

    total_messages: int = 0
    estimated_tokens: int = 0
    memory_count: int = 0
    file_context_count: int = 0
    reference_count: int = 0
    build_time_ms: float = 0.0


@dataclass
class ContextBuildResult:
    """上下文构建结果。

    Attributes:
        messages: 完整的消息列表（可直接发送给 LLM）。
        system_prompt: 系统 Prompt。
        memories: 记忆内容列表。
        unified_context: 统一上下文（场景、情感、时间等）。
        stats: 构建统计。
        status: 构建状态 (success/partial/failed)。
        errors: 错误详情。
    """

    messages: list[dict[str, str]] = field(default_factory=list)
    system_prompt: str = ""
    memories: list[str] = field(default_factory=list)
    unified_context: UnifiedContext | None = None
    stats: ContextStats = field(default_factory=ContextStats)
    status: str = "success"
    errors: list[dict[str, str]] = field(default_factory=list)


class UnifiedContextBuilder:
    """统一上下文构建器。

    对标 TS 侧 UnifiedContextBuilder。

    上下文系统的统一入口，整合所有上下文相关组件，提供简洁的构建接口。

    特性：
    - 单例模式：全局唯一实例
    - 缓存机制：系统 Prompt 缓存，减少重复构建
    - 错误隔离：每个组件独立 try-catch，不影响整体
    - 统计追踪：构建次数、耗时、缓存命中率
    """

    _instance: UnifiedContextBuilder | None = None

    def __init__(self) -> None:
        self._pipeline = UnifiedContextPipeline()
        self._llm_context_builder = LLMContextBuilder()

        self._cache_enabled: bool = True
        self._system_prompt_cache: str | None = None
        self._cache_timestamp: float = 0.0
        self._cache_ttl: float = 300.0

        self._build_count: int = 0
        self._total_build_time: float = 0.0
        self._cache_hits: int = 0
        self._cache_checks: int = 0

        self._orchestrator: Any = None

    @classmethod
    def get_instance(cls) -> UnifiedContextBuilder:
        """获取单例实例。

        Returns:
            UnifiedContextBuilder: 单例实例。
        """
        if cls._instance is None:
            cls._instance = cls()
        return cls._instance

    @classmethod
    def reset_instance(cls) -> None:
        """重置单例实例（测试用）。"""
        cls._instance = None

    def set_orchestrator(self, orchestrator: Any) -> None:
        """设置上下文编排器。

        Args:
            orchestrator: UnifiedContextOrchestrator 实例。
        """
        self._orchestrator = orchestrator

    def set_memory_engine(self, engine: Any) -> None:
        """设置记忆引擎。

        Args:
            engine: 记忆引擎实例。
        """
        self._pipeline.set_memory_engine(engine)

    def set_sovereignty_pipeline(self, pipeline: Any) -> None:
        """设置数据主权管道。

        Args:
            pipeline: 数据主权管道实例。
        """
        self._pipeline.set_sovereignty_pipeline(pipeline)

    async def build_context(
        self,
        user_input: str = "",
        user_id: str = "",
        include_system_prompt: bool = True,
        include_memory: bool = True,
        include_file_context: bool = False,
        resolve_references: bool = False,
        max_tokens: int = 4096,
        scene: str = "",
        history: list[dict[str, str]] | None = None,
        history_limit: int = 20,
    ) -> ContextBuildResult:
        """构建完整上下文（异步）。

        Args:
            user_input: 用户输入。
            user_id: 用户ID。
            include_system_prompt: 是否包含系统 Prompt。
            include_memory: 是否包含记忆。
            include_file_context: 是否包含文件上下文。
            resolve_references: 是否解析 @引用。
            max_tokens: 最大 Token 预算。
            scene: 场景类型。
            history: 历史消息。
            history_limit: 历史消息数量限制。

        Returns:
            ContextBuildResult: 构建结果。
        """
        start_time = time.time()
        self._build_count += 1

        options = ContextBuildOptions(
            user_input=user_input,
            user_id=user_id,
            include_system_prompt=include_system_prompt,
            include_memory=include_memory,
            include_file_context=include_file_context,
            resolve_references=resolve_references,
            max_tokens=max_tokens,
            scene=scene,
            history=history or [],
            history_limit=history_limit,
        )

        return await self._build(options, start_time)

    def build_context_sync(
        self,
        user_input: str = "",
        user_id: str = "",
        include_system_prompt: bool = True,
        include_memory: bool = True,
        include_file_context: bool = False,
        resolve_references: bool = False,
        max_tokens: int = 4096,
        scene: str = "",
        history: list[dict[str, str]] | None = None,
        history_limit: int = 20,
    ) -> ContextBuildResult:
        """构建完整上下文（同步）。"""
        start_time = time.time()
        self._build_count += 1

        options = ContextBuildOptions(
            user_input=user_input,
            user_id=user_id,
            include_system_prompt=include_system_prompt,
            include_memory=include_memory,
            include_file_context=include_file_context,
            resolve_references=resolve_references,
            max_tokens=max_tokens,
            scene=scene,
            history=history or [],
            history_limit=history_limit,
        )

        return self._build_sync(options, start_time)

    async def _build(
        self,
        options: ContextBuildOptions,
        start_time: float,
    ) -> ContextBuildResult:
        """异步执行构建。"""
        result = ContextBuildResult()
        errors: list[dict[str, str]] = []

        unified_context = await self._pipeline.build_context(
            user_input=options.user_input,
            user_id=options.user_id,
        )
        result.unified_context = unified_context

        if options.include_system_prompt:
            result.system_prompt = self._build_system_prompt(unified_context)

        if options.include_memory:
            memories = await self._build_memories(
                unified_context.memories,
                unified_context.scene.type,
                options.max_tokens,
            )
            result.memories = memories

        result.messages = self._build_messages(
            options=options,
            system_prompt=result.system_prompt,
            unified_context=unified_context,
        )

        result.stats = self._compute_stats(
            options=options,
            result=result,
            start_time=start_time,
        )

        if errors:
            result.status = "partial"
            result.errors = errors

        return result

    def _build_sync(
        self,
        options: ContextBuildOptions,
        start_time: float,
    ) -> ContextBuildResult:
        """同步执行构建。"""
        result = ContextBuildResult()
        errors: list[dict[str, str]] = []

        unified_context = self._pipeline.build_context_sync(
            user_input=options.user_input,
            user_id=options.user_id,
        )
        result.unified_context = unified_context

        if options.include_system_prompt:
            result.system_prompt = self._build_system_prompt(unified_context)

        if options.include_memory:
            memories = self._llm_context_builder.build_simple(
                unified_context.memories,
                scene=unified_context.scene.type,
            )
            result.memories = [m["content"] for m in memories]

        result.messages = self._build_messages(
            options=options,
            system_prompt=result.system_prompt,
            unified_context=unified_context,
        )

        result.stats = self._compute_stats(
            options=options,
            result=result,
            start_time=start_time,
        )

        if errors:
            result.status = "partial"
            result.errors = errors

        return result

    def _build_system_prompt(self, context: UnifiedContext) -> str:
        """构建系统 Prompt。

        Args:
            context: 统一上下文。

        Returns:
            系统 Prompt 字符串。
        """
        self._cache_checks += 1

        if self._cache_enabled and self._system_prompt_cache is not None:
            if time.time() - self._cache_timestamp < self._cache_ttl:
                self._cache_hits += 1
                return self._system_prompt_cache

        time_slot = context.time_context.time_slot
        scene = context.scene.type

        prompt_parts = [
            f"当前时间: {time_slot}",
            f"当前场景: {scene}",
        ]

        if context.emotion.type != "neutral":
            prompt_parts.append(
                f"用户情绪: {context.emotion.type} (强度: {context.emotion.intensity:.1f})"
            )

        prompt_parts.append("请根据当前场景和用户情绪提供合适的回应。")

        system_prompt = "\n".join(prompt_parts)

        if self._cache_enabled:
            self._system_prompt_cache = system_prompt
            self._cache_timestamp = time.time()

        return system_prompt

    async def _build_memories(
        self,
        raw_memories: list[dict[str, Any]],
        scene: str,
        max_tokens: int,
    ) -> list[str]:
        """异步构建记忆列表。

        Args:
            raw_memories: 原始记忆。
            scene: 场景类型。
            max_tokens: 最大 Token。

        Returns:
            记忆内容列表。
        """
        if not raw_memories:
            return []

        result = self._llm_context_builder.build(
            raw_memories,
            scene=scene,
            max_tokens=max_tokens,
        )

        return [m.content for m in result.memories]

    def _build_messages(
        self,
        options: ContextBuildOptions,
        system_prompt: str,
        unified_context: UnifiedContext,
    ) -> list[dict[str, str]]:
        """构建消息列表。

        Args:
            options: 构建选项。
            system_prompt: 系统 Prompt。
            unified_context: 统一上下文。

        Returns:
            消息列表。
        """
        messages: list[dict[str, str]] = []

        if system_prompt and options.include_system_prompt:
            messages.append({"role": "system", "content": system_prompt})

        if options.history:
            limited_history = options.history[-options.history_limit :]
            for msg in limited_history:
                messages.append({
                    "role": msg.get("role", "user"),
                    "content": msg.get("content", ""),
                })

        if options.user_input:
            messages.append({"role": "user", "content": options.user_input})

        return messages

    def _compute_stats(
        self,
        options: ContextBuildOptions,
        result: ContextBuildResult,
        start_time: float,
    ) -> ContextStats:
        """计算构建统计。

        Args:
            options: 构建选项。
            result: 构建结果。
            start_time: 开始时间。

        Returns:
            ContextStats: 构建统计。
        """
        build_time = (time.time() - start_time) * 1000
        self._total_build_time += build_time

        total_tokens = sum(len(m.get("content", "")) // 4 for m in result.messages)

        return ContextStats(
            total_messages=len(result.messages),
            estimated_tokens=total_tokens,
            memory_count=len(result.memories),
            file_context_count=0,
            reference_count=0,
            build_time_ms=build_time,
        )

    @property
    def build_count(self) -> int:
        """构建次数。"""
        return self._build_count

    @property
    def average_build_time_ms(self) -> float:
        """平均构建耗时。"""
        if self._build_count == 0:
            return 0.0
        return self._total_build_time / self._build_count

    @property
    def cache_hit_rate(self) -> float:
        """缓存命中率。"""
        if self._cache_checks == 0:
            return 0.0
        return self._cache_hits / self._cache_checks

    def clear_cache(self) -> None:
        """清除缓存。"""
        self._system_prompt_cache = None
        self._cache_timestamp = 0.0
        self._cache_hits = 0
        self._cache_checks = 0

    def reset_stats(self) -> None:
        """重置统计。"""
        self._build_count = 0
        self._total_build_time = 0.0
        self._cache_hits = 0
        self._cache_checks = 0
